MerlinAU.sh

```diff
@@ -5640,8 +5700,8 @@ _ChangelogVerificationCheck_()
             }')
 
             # Define regex patterns for both versions
-            release_version_regex="$formatted_release_version \([0-9]{1,2}-[A-Za-z]{3}-[0-9]{4}\)"
-            current_version_regex="$formatted_current_version \([0-9]{1,2}-[A-Za-z]{3}-[0-9]{4}\)"
+            release_version_regex="[0-9]{4}\.${formatted_release_version//./[._]}\s*\([0-9]{1,2}-[A-Za-z]+-[0-9]{4}\)"
+            current_version_regex="[0-9]{4}\.${formatted_current_version//./[._]}\s*\([0-9]{1,2}-[A-Za-z]+-[0-9]{4}\)"
```

Comment:
These regular expressions will not work for the "**Changelog-386.txt**" file because they're looking for 4 digits **before** the version string but they don't exist. You need to account for both cases: **with** or **without** the 4-digit prefix.

![ASUSWRT-Merlin_ChangeLog-386](https://github.com/user-attachments/assets/428984dd-76c6-4e6b-8435-46cf16877ef5)

Reply:
Ahhhh, good call. I didn't check that changelog, I only checked NG and 3006.

I'm assuming if I simply remove the requirement for the 3004 prefix. That it will work correctly? Could you test that?

Reply:
Just sitting back down from after dinner now. Give me a few moments to test the change or NG and 3006 as well

Reply:
> Ahhhh, good call. I didn't check that changelog, I only checked NG and 3006.
> 
> I'm assuming if I simply remove the requirement for the 3004 prefix. That it will work correctly? Could you test that?
>
>
Yeah, that's one way to do it. However, IMO, the better solution would be to fix the code above where all the 4-digit prefixes (i.e. "**3004**" & "**3006**") seem to get removed from the "formatted" version strings.

I think the key is to remove it **only** for the "**386**" builds but do not remove for the "**388**" & "**3006.102**" builds. Once you do that, the regular expression become much simpler.

My 2 cents.

Reply:
I think that was initially done exactly because of the 386 changelog. We only really needed to match the specific minor and patch version without caring about the prefix since the prefix is assumed within that specific changelog (3006, NG, etc)

Reply:
> Just sitting back down from after dinner now. Give me a few moments to test the change or NG and 3006 as well
>
I'll be "summoned for dinner" in about 10 minutes here... Just FYI. :>)

Reply:
The date regex matching is to make sure we don't match specific versions within the notes if I remember correctly.

Reply:
> I think that was initially done exactly because of the 386 changelog. We only really needed to match the specific minor and patch version without caring about the prefix since the prefix is assumed within that specific changelog (3006, NG, etc)
>
>
So previously the 4-digit prefix was kept **except** for the "386" builds?
If so, why was it changed? I don't see a good reason to have changed it...

Reply:
> The date regex matching is to make sure we don't match specific versions within the notes if I remember correctly.
>
Oh, yeah, there's nothing wrong with that part of the regex. We want to match **ONLY** the strings that have the versions PLUS the dates in one line.

Reply:
> > I think that was initially done exactly because of the 386 changelog. We only really needed to match the specific minor and patch version without caring about the prefix since the prefix is assumed within that specific changelog (3006, NG, etc)
> 
> So previously the 4-digit prefix was kept **except** for the "386" builds? If so, why was it changed? I don't see a good reason to have changed it...

No I'm saying the code above removes the prefix since the beginning. It was "initially" done as in the original design removed the prefix for that 386 changelog. It's not included in the changelog. So we removed it from the current and upcoming version strings for comparison against the changelogs.

The prefix is not required for the 3004 or 3006 changelogs since the prefix is assumed already within those changelogs.

Reply:
> The "**3004**" & "**3006**" prefixes seem to get removed by the `awk` code above even though the specific prefix **is** **required** for the "**388**" & "**102**" builds to search for the "version + date" in the changelog.
> 

Is it required though? My understanding of the fix is me removing the suffix from the code above, the regex was just being adjusted to be more forgiving with the _ or . And the spacing around the date. I added the prefix because I only checked NG and 3006 but I don't think it's required if wasn't in the code previously.

Reply:
In my test here: it doesn't appear to be required:

![image](https://github.com/user-attachments/assets/de08e59d-3294-413c-8798-76e18ed153fc)

Reply:
Checkout commit: https://github.com/ExtremeFiretop/MerlinAutoUpdate-Router/pull/367/commits/2c70b4c40bd2223bdd85d804f235ce46ae59dd62

That was my suggestion, I don't think the prefix is required in those changelogs for results to be matched.

Reply:
Btw I'm not trying to be dismissive of your idea to modify the top code above the regex to only drop the prefix for 386. 

I'm just trying to understand why keep the prefix at all if it's not required for the matching to work?

I only initially added the prefix match to the regex because I made the regex more forgiving and loser matching with the rear of the version (swapping dots or underscores) so I decided to tighten up the front end of the version number to balance out the change and not make the regex too lose. 

But really the date matching is good enough and the prefix doesn't appear to be required so it can just be dropped from the regex and now it works with all changelogs again.

The two real fixes here are the dropping of the suffix in the code above for the current version, and making the regex loser with the dots and underscore matching since some changelogs use _2 and others might use .2

Reply:
> Btw I'm not trying to be dismissive of your idea to modify the top code above the regex to only drop the prefix for 386.
> 
> I'm just trying to understand why keep the prefix at all if it's not required for the matching to work?
> 
> I only initially added the prefix match to the regex because I made the regex more forgiving and loser matching with the rear of the version (swapping dots or underscores) so I decided to tighten up the front end of the version number to balance out the change and not make the regex too lose.
> 
> But really the date matching is good enough and the prefix doesn't appear to be required so it can just be dropped from the regex and now it works with all changelogs again.
> 
> The two real fixes here are the dropping of the suffix in the code above for the current version, and making the regex loser with the dots and underscore matching since some changelogs use _2 and others might use .2 and being more forgiving on the amount of spacing between the date and the version.
>
>

I believe this is a case where we have differences of opinion WRT "proper implementation" because we're operating under a different set of coding rules & standards. Let me explain a little. Pretty much all rules & standards I've worked with dictate that whenever we're trying to find a matching line, we always try to search for as much of the expected string as possible, from the beginning to the end, and make exceptions to the rule only for very special circumstances.

For the particular case of the changelogs, my default operating mode is to match as much of the specific version string as possible followed by the date string. This means that I would try to match the full version, including the 4-digit prefix (i.e. "3004" or "3006") unless there's a special case (i.e. "386" builds). Therefore, in my view, the 4-digit prefixes are required for proper line matching, taking into account exceptions for special case(s), because we're looking for precise matches.

Again, this is how I always operate in my professional work, and using these coding rules & standards for my implementations spills over to all extracurricular projects & hobbies I collaborate in.

Now, I know there are "many ways to skin a cat." And we obviously have different coding backgrounds & standards, so I understand your point of view even though I would have done it differently to match my own working operating mode.

I hope the above clarifies where I'm coming from WRT my coding recommendations for solutions.

Ultimately, if your fixes are working well for all test cases, then we're good to go.

Reply:
Yes that clarifies what you mean by it being required. Because in your standards it is and I'm not familiar with that.

But at a technical level when I tested the fix it didn't seem to physically be required which is why the original implementation from the start dropped the prefixes.

Now I think I'm understanding you want to change it so everything uses the prefixes by default? And only for builds detected as 386 do we do the matching by dropping the prefix?

I think I understand the viewpoint, I just was tagging a fix with minimal changes required, and already I overstepped by adding the prefix match and not thinking / double checking the 386 changelogs.

Reply:
> Yes that clarifies what you mean by it being required. Because in your standards it is and I'm not familiar with that.
> 
> But at a technical level when I tested the fix it didn't seem to physically be required which is why the original implementation from the start dropped the prefixes.
> 
> Now I think I'm understanding you want to change it so everything uses the prefixes by default? And only for builds detected as 386 do we do the matching by dropping the prefix?
> 
> I think I understand the viewpoint, I just was tagging/thinking of a fix with minimal changes required, and already I overstepped by adding the prefix match and not thinking / double checking the 386 changelogs.]
>
No worries, we're good to go for this round of fixes. If in the future any bugs are found in this area again, we can revisit the implementation.

Let me connect to my company's server and then my workstation to run the Linter tool. I'll let you know of the results.